prql-compiler/src/lib.rs

```diff
@@ -22,2218 +57,56 @@ static PRQL_VERSION: Lazy<Version> =
 
 /// Compile a PRQL string into a SQL string.
 ///
-/// This has three stages:
-/// - [parse] — Build PL AST from a PRQL string
-/// - [semantic::resolve] — Finds variable references, validates functions calls, determines frames and converts PL to RQ.
-/// - [sql::translate] — Convert RQ AST into an SQL string.
-pub fn compile(prql: &str) -> Result<String> {
-    parse(prql).and_then(semantic::resolve).and_then(translate)
+/// This is a wrapper for:
+/// - [prql_to_pl] — Build PL AST from a PRQL string
+/// - [pl_to_rq] — Finds variable references, validates functions calls, determines frames and converts PL to RQ.
+/// - [rq_to_sql] — Convert RQ AST into an SQL string.
+pub fn compile(prql: &str) -> Result<String, ErrorMessage> {
+    parser::parse(prql)
+        .and_then(semantic::resolve)
+        .and_then(|rq| sql::compile(rq, None))
+        .map_err(error::downcast)
+        .map_err(|e| e.composed("", prql, false))
 }
 
-/// Generate PRQL code from PL AST
-pub fn pl_to_prql(pl: Vec<Stmt>) -> Result<String> {
-    Ok(format!("{}", ast::pl::Statements(pl)))
+/// Parse PRQL into a PL AST
+pub fn prql_to_pl(prql: &str) -> Result<Vec<Stmt>, ErrorMessage> {
+    parser::parse(prql)
+        .map_err(error::downcast)
+        .map_err(|e| e.composed("", prql, false))
 }
 
-/// Serialize PL AST into JSON
-pub fn pl_to_json(pl: Vec<Stmt>) -> Result<String> {
-    Ok(serde_json::to_string(&pl)?)
+/// Perform semantic analysis and convert PL to RQ.
+pub fn pl_to_rq(pl: Vec<Stmt>) -> Result<Query, ErrorMessage> {
```

Comment:
These are both conceptually elegant, like the previous iteration, and easy to find 👏